### ISSUE TITLE
Code completion breaks server with macro based class

### ISSUE THREAD

**Issue by hoseyjoe** [vshaxe/vshaxe#349]
I have a class that is created from xml using a macro.   I can compile it multiple times using command line or vscode with  
haxe.exe --connect 6000 html5.hxml

If I open the file that references the class creation  with line like
var myclass=createclass("Assets/Main.xml");
I start getting an error
Type not found : Assets.Main
which is the dynamic class name. At which point the cache is broken and calling  
haxe.exe --connect 6000 html5.hxml
no longer works. it throws the same error. Closing the file and restarting language server fixes issue.

Since it compiles with language server I am assuming it is something on vshaxe side??

**Comment by Gama11** [comment]
Please provide a code example with which the problem can be reproduced.

**Comment by hoseyjoe** [comment]
ya, I had a feeling I should have created a sample project first. You need to be using new-component-method branch of both haxeui-core and haxeui-html5 for it to compile. then open Main.hx and you should start getting

Type not found : assets.Main  

src/Main.hx:8: lines 8-22 : Defined in this class  

The terminal process terminated with exit code: 1

[_master.zip](https://github.com/vshaxe/vshaxe/files/3228813/_master.zip)

**Comment by Gama11** [comment]
Thanks for the test project, that's helpful.

This seems to be caused by diagnostics running on `Main.hx` somehow, so `"haxe.enableDiagnostics": false` is a potential workaround for now.

**Comment by RealyUniqueName** [comment]
I cannot build that sample.
With Haxe 3.4.7:
```
$ haxelib install html5.hxml
<...>
$ haxe html5.hxml 
src/Main.hx:12: characters 23-53 : Class<haxe.ui.macros.ComponentMacros> has no field createInstance
src/Main.hx:16: characters 29-33 : {+ btn : {+ onClick : e : Unknown<0> -> Void } } should be haxe.ui.core.Component
src/Main.hx:16: characters 29-33 : For function argument 'component'
```
With the latest dev Haxe:
```
$ haxe html5.hxml 
/home/alex/temp/_master/.haxelib/haxeui-core/0,0,4/haxe/ui/Toolkit.hx:43: characters 10-11 : Expected }
/home/alex/temp/_master/.haxelib/haxeui-core/0,0,4/haxe/ui/Toolkit.hx:267: characters 14381-14382 : Expected }
/home/alex/temp/_master/.haxelib/haxeui-core/0,0,4/haxe/ui/Toolkit.hx:24: lines 24-267 : Defined in this class
/home/alex/temp/_master/.haxelib/haxeui-core/0,0,4/haxe/ui/Toolkit.hx:267: characters 3019-3020 : Expected }
/home/alex/temp/_master/.haxelib/haxeui-core/0,0,4/haxe/ui/Toolkit.hx:24: lines 24-267 : Defined in this class
src/Main.hx:12: characters 24-54 : Class<haxe.ui.macros.ComponentMacros> has no field createInstance
src/Main.hx:16: characters 30-34 : {+ btn : {+ onClick : (e : Unknown<0>) -> Void } } should be haxe.ui.core.Component
src/Main.hx:16: characters 30-34 : For function argument 'component'
```

**Comment by Gama11** [comment]
You need haxeui from GitHub, the Haxelib releases are quite outdated. To quote @hoseyjoe:

>You need to be using new-component-method branch of both haxeui-core and haxeui-html5 for it to compile.

**Comment by RealyUniqueName** [comment]
Ok. It builds. But once I add `var myclass=createclass("Assets/Main.xml");` I get:
```
src/Main.hx:19: characters 25-36 : Unknown identifier : createclass
```

**Comment by Simn** [comment]
The example doesn't work for me due to `Cannot use abstract as value` and `Cannot use Void as value` errors. However, if I disable those in Haxe I can get completion in that Main.hx.

**Comment by RealyUniqueName** [comment]
Is this issue still relevant?

**Comment by ianharrigan** [comment]
This still happens for me (late to the party, i know). Could it be that the language server doesnt have the file "assets/main.xml", and therefore cant build the assets.Main class?

I might be way off with that assumption, but if im not, is there a way to tell the language server "here is a file"?

Cheers,
Ian

**Comment by ianharrigan** [comment]
Ok, after a little investigation this seems to be fixed now in `haxeui-core:1.0.9`. I dont quite understand the why its needed, but it seems i needed to clear the `ComponentClassMap` when the module macros run. 

I looks _something_ like when the first compile happens (building cache for vscode) it generates the dynamic class and adds its class name to the `ComponentClassMap`, then on subsequent compiles its goes to generate it again, but finds its already in the `ComponentClassMap` so skips it, however, it doesnt seem it has actually been generated, so compilation fails. 

Clearing the class map fixes that by making sure its always (re)generated. Im not sure if its a workaround or a bugfix in haxeui, but it seems to function, and, apart from having to gen the class again, isnt a big deal imo - @hoseyjoe, that work for you?

Cheers,
Ian

**Comment by Gama11** [comment]
Closing this as it seems to have been fixed / worked around in haxeui, and in any case wouldn't be a vshaxe but a Haxe issue.
